lib/spack/spack/build_environment.py

```diff
@@ -321,12 +321,32 @@ def set_build_environment_variables(pkg, env, dirty=False):
         extra_rpaths = ':'.join(compiler.extra_rpaths)
         env.set('SPACK_COMPILER_EXTRA_RPATHS', extra_rpaths)
 
-    # Add bin directories from dependencies to the PATH for the build.
-    bin_dirs = reversed(filter(os.path.isdir, [
-        '%s/bin' % d.prefix for d in pkg.spec.dependencies(deptype='build')]))
-    bin_dirs = filter_system_bin_paths(bin_dirs)
-    for item in bin_dirs:
-        env.prepend_path('PATH', item)
+    # Add bin directories from dependencies to the PATH for the build or run.
+    def add_dir_to_list(bin_dir, list):
+        if os.path.isdir(bin_dir):
+            if bin_dir not in list:
+                list.append(bin_dir)
+    build_deps = [ d for d in pkg.spec.dependencies(deptype='build') ]
+    build_bin_dirs = []
```

Comment:
could you add a short comment above `build_bin_dirs = []`, `run_bin_dirs = []` and `bin_dirs = []` what those are meant to contain.